modules/nixos/tx5-signal-server.nix

```diff
@@ -0,0 +1,119 @@
+{
+  self,
+  config,
+  lib,
+  pkgs,
+  ...
+}: let
+  cfg = config.services.tx5-signal-server;
+in {
+  options.services.tx5-signal-server = {
+    enable = lib.mkEnableOption "tx5-signal-server";
+
+    package = lib.mkOption {
+      default = self.packages.${pkgs.system}.tx5-signal-srv;
+      type = lib.types.package;
+    };
+
+    address = lib.mkOption {
+      description = "address to bind";
+      type = lib.types.str;
+    };
+
+    tls-port = lib.mkOption {
+      description = "port to bind for incoming TLS connections";
+      type = lib.types.int;
+    };
+
+    url = lib.mkOption {
+      description = "url for incoming TLS connections to the signal server";
+      type = lib.types.str;
+    };
+
+    port = lib.mkOption {
+      description = "port to bind";
+      type = lib.types.int;
+    };
+
+    iceServers = lib.mkOption {
+      description = "webrtc configuration to broadcast";
+      type = lib.types.listOf lib.types.attrs;
+      default = [];
+    };
+
+    demo = lib.mkEnableOption "enable demo broadcasting as a stand-in for bootstrapping";
+
+    configTextFile = lib.mkOption {
+      default = builtins.toFile "tx5.config.json" (builtins.toJSON {
+        port = cfg.port;
+        iceServers.iceServers = cfg.iceServers;
+        demo = cfg.demo;
+      });
+    };
+  };
+
+  config = lib.mkIf (cfg.enable) {
+    systemd.services.tx5-signal-server = {
+      after = ["network.target"];
+      wantedBy = ["multi-user.target"];
+
+      environment = {
+        TMPDIR = "%T";
+        CONFIG_PATH = "%T/config.json";
+      };
+
+      serviceConfig = {
+        DynamicUser = true;
+        PrivateTmp = true;
+        ExecStartPre = pkgs.writeShellScript "tx5-start-pre" ''
```

Comment:
Is it not possible to put this in a directory without scripting? Like `/var/lib/bootstrap/config.json` or somewhere like that. Not really bothered where but I think there's a way to output files to standard locations to be consumed by services with NixOS?

Reply:
nix can by design only put files in the nix store. the point here isn't that it's not in the nix store though, it's that the permissions are `0400` which is enforced by the `tx5-signal-srv`.

Reply:
in addition i also thought about how we can inject secrets into the config.json at runtime. it turns out we're not needing that now because we've got a hashed password in the config.

Reply:
I think you can create files alongside services like this https://unix.stackexchange.com/questions/500025/how-to-add-a-file-to-etc-in-nixos

Reply:
> I think you can create files alongside services like this https://unix.stackexchange.com/questions/500025/how-to-add-a-file-to-etc-in-nixos

using `environment.etc`, the file is going to be written to the nix store first, and then linked to from `/etc` with the link's permissions being adjusted. the file in the nix store will we world-readable which i'd like to avoid in this case. that's because i want to reserve the functionality of injecting secret information into this config, which should never reach the nix store in plain text.

what do you think are the downsides of the current implementation?

Reply:
Ahh okay, got it, yeah that's fine then